File: src/lifetimes.ts
// What if our Point struct wanted to store *references* instead?
export type Point<T> = {
    timestamp: Date,
    value: T, // We now require a lifetime specifier.
}

export type ISeries = {
    points?: Point<number>[],
}

class Series implements ISeries {
    constructor(public points?: Point<number>[]) { }

    from_points(points: Point<number>[]): Series {
        return new Series(points);
    }

    toString(): string {
        return `Series { points: ${JSON.stringify(this.points)} }`;
    }
}

/// # Output
let series = new Series([
    { timestamp: new Date(), value: 1 } as Point<number>,
    { timestamp: new Date(), value: 2 } as Point<number>,
])

console.log(`🐥 series: ${series}`);

// 🐥 series: Series { points: [{"timestamp":"2022-05-09T04:09:28.831Z","value":1},{"timestamp":"2022-05-09T04:09:28.831Z","value":2}] }

File: src/lifetimes.rs
// What if our Point struct wanted to store *references* instead?
#[derive(Clone, Debug)]
pub struct Point<T> {
    timestamp: time::PrimitiveDateTime,
    value: T, // We now require a lifetime specifier.
}

#[derive(Debug)]
pub struct Series {
    points: Vec<Point<f64>>,
}

impl Series {
    pub fn new() -> Self {
        Self { points: Vec::new() }
    }

    pub fn from_points(points: &[Point<f64>]) -> Self {
        Self {
            points: points.to_owned(),
        }
    }
}

/// # Output
#[test]
fn test() {
    use time::macros::datetime;
    let series = Series::from_points(&[
        Point {
            timestamp: datetime!(2019-01-01 0:00),
            value: 1.0,
        },
        Point {
            timestamp: datetime!(2019-01-02 0:00),
            value: 2.0,
        },
    ]);
    println!("🦀 series: {:?}", series);
}

// 🦀 series: Series { points: [Point { timestamp: PrimitiveDateTime { date: Date { year: 2019, ordinal: 1 }, time: Time { hour: 0, minute: 0, second: 0, nanosecond: 0 } }, value: 1.0 }, Point { timestamp: PrimitiveDateTime { date: Date { year: 2019, ordinal: 2 }, time: Time { hour: 0, minute: 0, second: 0, nanosecond: 0 } }, value: 2.0 }] }
